processor/src/main/java/com/linecorp/decaton/processor/processors/BatchingProcessor.java

```diff
@@ -0,0 +1,145 @@
+            this.completion = completion;
+            this.context = context;
+            this.task = task;
+        }
+    }
+
+    /**
+     * Instantiate {@link BatchingProcessor}.
+     * If you only need one limit, please set large enough value to another.
+     * @param lingerMillis time limit for this processor. On every lingerMillis milliseconds,
+     * tasks in past lingerMillis milliseconds are pushed to {@link BatchingTask#processBatchingTasks(List)}.
+     * @param capacity size limit for this processor. Every time tasks’size reaches capacity,
+     * tasks in past before reaching capacity are pushed to {@link BatchingTask#processBatchingTasks(List)}.
+     */
+    protected BatchingProcessor(long lingerMillis, int capacity) {
+        this.lingerMillis = lingerMillis;
+        this.capacity = capacity;
+
+        ScheduledThreadPoolExecutor scheduledExecutor = new ScheduledThreadPoolExecutor(1, r -> {
+            Thread th = new Thread(r);
+            th.setName("Decaton" + BatchingProcessor.class.getSimpleName()
+                       + '/' + System.identityHashCode(this));
+            return th;
+        });
+
+        // For faster shutdown cancel all pending flush on executor shutdown.
+        // In fact for this purpose we have two options here:
+        // A. Cancel all pending flush
+        // B. Run all pending flush immediately, regardless remaining time until flush
+        // A has a downside that it forces consumer to re-process some amount of tasks.
+        // B has a downside that it may causes flush bursting and then on spike for storage/API access in
+        // downstream processor.
+        // Given that this feature is expected to be used mainly for reducing workload for storage/APIs,
+        // we think taking A is much desirable behavior in most cases.
+        scheduledExecutor.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+        executor = scheduledExecutor;
+
+        scheduleFlush();
+    }
+
+    private void flush(List<BatchingTask<T>> batchingTasks) {
+        if (batchingTasks.isEmpty()) {
+            return;
+        }
+        processBatchingTasks(batchingTasks);
+    }
+
+    private Runnable periodicallyFlushTask() {
+        return () -> {
+            synchronized (windowedTasks) {
+                flush(new ArrayList<>(windowedTasks));
+                windowedTasks.clear();
+            }
+            scheduleFlush();
+        };
+    }
+
+    // visible for testing
+    protected void scheduleFlush() {
+        executor.schedule(periodicallyFlushTask(), lingerMillis, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void process(ProcessingContext<T> context, T task) throws InterruptedException {
+        BatchingTask<T> newTask = new BatchingTask<>(context.deferCompletion(), context, task);
+        windowedTasks.add(newTask);
+        if (windowedTasks.size() >= this.capacity) {
```

Comment:
>I think this should be executed in scheduler thread to avoid unnecessary lock contention between Decaton's processor thread and scheduler thread

Thank you! I finally understood the meaning of the above sentence.
`executor.submit(() -> processBatchingTasks(batch));`  is important.
I fixed it. 9e35f1b